### DISCUSSION TITLE
What i (and GPT) did for enhancing Audit log readability with dynamic property translation

### DISCUSSION THREAD

**Discussion by malkafly** [owen-it/laravel-auditing#900] [Ideas]
Hello everyone,

I recently embarked on a mission to enhance the readability of audit logs generated by our beloved Laravel Auditing package. The objective was to transform the cryptic property names in the `new_values` and `old_values` JSON columns into human-readable titles, making the audit logs more accessible and easier to understand.

With the assistance of ChatGPT, I crafted a solution that not only improves the log readability but also enriches the overall user experience when dealing with audit data. The process began with extracting all unique property names directly from the audit logs, which involved querying the `audits` table and dynamically listing all column names involved in the `new_values` and `old_values`, excluding common Laravel timestamps and IDs.

Here's a snippet of how the unique property names were extracted:

```php
$auditLogs = Audit::all(); // Assuming Audit is your audit log model
$uniqueProperties = [];

foreach ($auditLogs as $log) {
    $newValues = json_decode($log->new_values, true) ?? [];
    $oldValues = json_decode($log->old_values, true) ?? [];
    $properties = array_merge(array_keys($newValues), array_keys($oldValues));
    $uniqueProperties = array_unique(array_merge($uniqueProperties, $properties));
}

// Filter out common Laravel columns
$filteredProperties = array_diff($uniqueProperties, ['id', 'created_at', 'updated_at', 'deleted_at']);
```

After identifying the unique properties, I manually mapped them to more descriptive titles. ChatGPT was instrumental in generating a comprehensive dictionary that translates technical column names into user-friendly terms. Here’s a glimpse of the translation array and how it's utilized:

```php
class Audit extends Model
{
    public static $propertyTitles = [
        "acessado" => "Accessed",
        "acompanhante" => "Companion",
        "acompanhante_contato" => "Companion Contact",
        // Additional mappings
    ];

    public static function translateProperty($key)
    {
        return static::$propertyTitles[$key] ?? $key; // Fallback to the key if no translation is available
    }
}
```

And in the Blade template, making the audit data presentation more intuitive:

```blade.php

<table>
    <thead>
        <tr>
            <th>Campo</th>
            <th>Dados Antigos</th>
            <th>Dados Novos</th>
        </tr>
    </thead>
    <tbody>
        @foreach ($data as $item)
            @php
                $oldValues = json_decode($item->old_values, true) ?? [];
                $newValues = json_decode($item->new_values, true) ?? [];
                $allFields = array_unique(array_merge(array_keys($oldValues), array_keys($newValues)));
            @endphp

            @foreach ($allFields as $field)
                <tr>
                    <td>{{ \App\Models\Audit::translateProperty($field) }}</td>
                    <td>{{ array_key_exists($field, $oldValues) ? $oldValues[$field] : '' }}</td>
                    <td>{{ array_key_exists($field, $newValues) ? $newValues[$field] : '' }}</td>
                </tr>
            @endforeach
        @endforeach
    </tbody>
</table>
```

This enhancement significantly improved the audit log readability and overall user experience. I am excited to share this idea with the community and hear your thoughts. Perhaps it could inspire further enhancements or even be integrated into the package.

Result:
![image](https://github.com/owen-it/laravel-auditing/assets/233273/76caf0f2-65db-4075-ace1-42684249765e)

> Perhaps there is a package/component or a better way to accomplish this, but I was bored and had GPT open, so I couldn't stop until I finished. If there's a smarter and quicker method, this thread would be great for that. Thanks!

**Comment by parallels999** [comment]
`$propertyTitles` shouldn't it be in the audited model? _On this case, no need to change the package_
If it were in the `Audit` model, would this property grow too much and would it be mixing all the models?

```php
\App\Models\MyAuditableModel::translateProperty($field)
```
